src/pudl/transform/ferc1.py

```diff
@@ -97,6 +101,138 @@ class Ferc1RenameColumns(TransformParams):
+
+    :func:`wide_to_tidy_xbrl` will generate a regex pattern assuming the ``value_types``
+    are the column name's suffixes. If a column does not conform to that pattern, it
+    will be filtered out. This is helpful for us to not include a bunch of columns from
+    the input dataframe incorrectly included in the stacking process. We could enumerate
+    every column that we want to drop, but this could be tedious and potentially error
+    prone. But this does mean that if a column is incorrectly named - or barely missing
+    the pattern, it will be dropped. This parameter enables us to lock the number of
+    expected columns. If the dropped columns are a different number, an error will be
+    raised.
+    """
+
+
+def wide_to_tidy_xbrl(df: pd.DataFrame, params: WideToTidyXBRL) -> pd.DataFrame:
+    """Reshape wide tables with FERC account columns to tidy format.
+
+    The XBRL table coming into this method contains all the data from both the
+    instant and duration tables in a wide format -- with one column for every
+    combination of value type (e.g. additions, ending_balance) and accounting
+    category, which means ~500 columns.
+
+    We tidy this into a long table with one column for each of the value types (6 in
+    all), and a new column that contains the accounting categories. This allows
+    aggregation across columns to calculate the ending balance based on the starting
+    balance and all of the reported changes, and aggregation across groups of rows
+    to total up various hierarchical accounting categories (hydraulic turbines ->
+    hydraulic production plant -> all production plant -> all electric utility
+    plant) though the categorical columns required for that aggregation are added
+    later.
+    """
+    suffixes = "|".join(params.value_types)
+    pat = r"(^.*)_(" + suffixes + r"$)"
+    # filter out any columns that don't match the pattern
+    df_out = df.set_index(params.idx_cols).filter(regex=pat)
+    # check if there are unexpected columns being dropped
+    dropped_cols = [col for col in df if col not in df_out.reset_index()]
+    logger.debug(f"dropping: {dropped_cols}")
+    if params.expected_drop_cols != len(dropped_cols):
+        raise AssertionError(
+            f"Dropping more columns ({len(dropped_cols)}) than expected "
+            f"({params.expected_drop_cols}). Columns dropped: {dropped_cols}"
+        )
+
+    new_cols = pd.MultiIndex.from_tuples(
+        [(re.sub(pat, r"\1", col), re.sub(pat, r"\2", col)) for col in df_out.columns],
+        names=["xbrl_column_stem", "value_type"],
+    )
+    df_out.columns = new_cols
+    df_out = (
+        df_out.stack(level="xbrl_column_stem", dropna=False)
+        .loc[:, params.value_types]
+        .reset_index()
+    )
+    # remove the name of the columns which was the name of the renaming layer of the
+    # multi-index
+    df_out.columns.name = None
+    return df_out
+
+
+class MergeMetadata(TransformParams):
```

Comment:
Can we make it explicit that we're merging XBRL metadata with the param & function names? There are a lot of different kinds of metadata floating around.

Reply:
very good point